Generation of events for ServiceComm 0.11 and 0.12 fails

When generating an IDL file with events such as BTC.GA.FieldLink.idl, the generated code targeting ServiceComm 0.11 or 0.12 does not compile. It fails with an error like
```
[ERROR-] D:\NLS\serviceidl-integrationtests\output-263\cpp\modules/Archive/Proxy/include/CArchiveAPIProxy.h(22): fatal error C1083: Cannot open include file: 'ServiceComm/Commons/include/CSharedMessage.h': No such file or directory [D:\NLS\serviceidl-integrationtests\output-263\cpp\build\B\5\BTC.GA.Archive.Proxy.vcxproj]
```

--- a/com.btc.serviceidl.tests/src/com/btc/serviceidl/tests/generator/IdlGeneratorTest.xtend
+++ b/com.btc.serviceidl.tests/src/com/btc/serviceidl/tests/generator/IdlGeneratorTest.xtend
@@ -60,4 +60,17 @@ class IdlGeneratorTest
         println(fsa.textFiles.keySet.join("\n"))
         assertEquals(139, fsa.textFiles.size)
     }
+
+    @Test
+    def void testEvent()
+    {
+        val defaultGenerationSettingsProvider = generationSettingsProvider as DefaultGenerationSettingsProvider
+        defaultGenerationSettingsProvider.reset() // TODO remove this, it is necessary because the dependencies are reused across test cases        
+        val spec = TestData.eventTestCase.parse
+        val fsa = new InMemoryFileSystemAccess()
+        val generatorContext = new GeneratorContext()
+        underTest.doGenerate(spec.eResource, fsa, generatorContext)
+        println(fsa.textFiles.keySet.join("\n"))
+        assertEquals(117, fsa.textFiles.size)
+    }
 }

--- a/com.btc.serviceidl.tests/src/com/btc/serviceidl/tests/testdata/TestData.xtend
+++ b/com.btc.serviceidl.tests/src/com/btc/serviceidl/tests/testdata/TestData.xtend
@@ -34,6 +34,7 @@ class TestData
 
     }
 
+    // TODO rename or change this, it only uses a small set of features of the IDL
     static def CharSequence getFull()
     {
         return '''
@@ -60,6 +61,23 @@ class TestData
             }
             }
             }
+            }
+        '''
+
+    }
+
+    static def CharSequence getEventTestCase()
+    {
+        return '''
+            module foo {
+                
+            interface Test[version=1.0.0 guid=384E277A-C343-4F37-B910-C2CE6B37FC8E] {
+                event TestEvent [guid = 9BFE9AB7-3AA6-441F-82D7-D1A27F6321CE] (TestEventArgs);
+                
+                struct TestEventArgs{
+                  string text;
+                };
+            };
             }
         '''
 

--- a/com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/BasicCppGenerator.xtend
+++ b/com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/BasicCppGenerator.xtend
@@ -447,4 +447,9 @@ class BasicCppGenerator
         '''
     }
 
+    protected def getTargetVersion()
+    {
+        targetVersionProvider.getTargetVersion(CppConstants.SERVICECOMM_VERSION_KIND)
+    }
+
 }

--- a/com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/DispatcherGenerator.xtend
+++ b/com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/DispatcherGenerator.xtend
@@ -247,11 +247,6 @@ class DispatcherGenerator extends BasicCppGenerator
         '''
     }
 
-    private def getTargetVersion()
-    {
-        targetVersionProvider.getTargetVersion(CppConstants.SERVICECOMM_VERSION_KIND)
-    }
-
     private def getClientIdentityType()
     {
         if (targetVersion.equals("0.10"))

--- a/com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/HeaderResolver.xtend
+++ b/com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/HeaderResolver.xtend
@@ -104,6 +104,7 @@ class HeaderResolver
         "BTC::ServiceComm::API::EventKind" -> "ServiceComm/API/include/IEventRegistry.h",
         "BTC::ServiceComm::API::IClientEndpoint" -> "ServiceComm/API/include/IClientEndpoint.h",
         "BTC::ServiceComm::API::IEventSubscriberManager" -> "ServiceComm/API/include/IEventRegistry.h",
+        "BTC::ServiceComm::API::IEventSubscriberManager::ObserverType::OnNextParamType" -> "ServiceComm/API/include/IEventRegistry.h",
         "BTC::ServiceComm::API::InvalidMessageReceivedException" -> "ServiceComm/API/include/ServiceHostException.h",
         "BTC::ServiceComm::API::InvalidRequestReceivedException" -> "ServiceComm/API/include/ServiceHostException.h",
         "BTC::ServiceComm::API::IServerEndpoint" -> "ServiceComm/API/include/IServerEndpoint.h",

--- a/com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/ProxyGenerator.xtend
+++ b/com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/ProxyGenerator.xtend
@@ -70,9 +70,11 @@ class ProxyGenerator extends BasicCppGenerator {
             {
                «/* TODO remove ProtobufType argument */»
                «typeResolver.resolveProtobuf(event.data, ProtobufType.REQUEST)» eventProtobuf;
+               «IF targetVersion == "0.10"»
                if (!(event->GetNumElements() == 1))
                   «resolveSymbol("CABTHROW_V2")»(«resolveSymbol("BTC::ServiceComm::API::InvalidMessageReceivedException")»("Event message has not exactly one part"));
-               «resolveSymbol("BTC::ServiceComm::ProtobufUtil::ProtobufSupport")»::ParseMessageOrThrow<«resolveSymbol("BTC::ServiceComm::API::InvalidMessageReceivedException")»>(eventProtobuf, (*event)[0]);
+               «ENDIF»
+               «resolveSymbol("BTC::ServiceComm::ProtobufUtil::ProtobufSupport")»::ParseMessageOrThrow<«resolveSymbol("BTC::ServiceComm::API::InvalidMessageReceivedException")»>(eventProtobuf, «IF targetVersion == "0.10"»(*event)[0]«ELSE»*event«ENDIF»);
 
                return «typeResolver.resolveCodecNS(paramBundle, event.data)»::Decode( eventProtobuf );
             }
@@ -95,7 +97,7 @@ class ProxyGenerator extends BasicCppGenerator {
            return «resolveSymbol("CABTYPENAME")»(«event_type»);
          }
          
-         «resolveSymbol("std::function")»<«class_name.shortName»::«event_params_name»::EventDataType const ( «resolveSymbol("BTC::ServiceComm::Commons::ConstSharedMessageSharedPtr")» const & )> «class_name»::«event_params_name»::GetUnmarshalFunction( )
+         «resolveSymbol("std::function")»<«class_name.shortName»::«event_params_name»::EventDataType const ( «resolveSymbol("BTC::ServiceComm::API::IEventSubscriberManager::ObserverType::OnNextParamType")»)> «class_name»::«event_params_name»::GetUnmarshalFunction( )
          {
            return &Unmarshal«event_name»;
          }

--- a/com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/ServiceAPIGenerator.xtend
+++ b/com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/ServiceAPIGenerator.xtend
@@ -143,7 +143,7 @@ class ServiceAPIGenerator extends BasicCppGenerator {
                      static «resolveSymbol("BTC::Commons::CoreExtras::UUID")» GetEventTypeGuid();
                      static «resolveSymbol("BTC::ServiceComm::API::EventKind")» GetEventKind();
                      static «resolveSymbol("BTC::Commons::Core::String")» GetEventTypeDescription();
-                     static «resolveSymbol("std::function")»<EventDataType const ( «resolveSymbol("BTC::ServiceComm::Commons::ConstSharedMessageSharedPtr")» const & )> GetUnmarshalFunction();
+                     static «resolveSymbol("std::function")»<EventDataType const ( «resolveSymbol("BTC::ServiceComm::API::IEventSubscriberManager::ObserverType::OnNextParamType")» )> GetUnmarshalFunction();
                   };
                   «resolveSymbol("BTC::ServiceComm::Util::CDefaultObservableRegistrationProxy")»<«event_params_name»> «event.observableRegistrationName»;
                «ENDFOR»

